### PR TITLE
WT-9501 Implement functionality to test cursor search in cursor_bound_01

### PR DESCRIPTION
**Pull request by clarissecheah**

### CODE REVIEW

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 1]
```diff
@@ -289,6 +289,50 @@ class cursor_bound_01 : public test {
 
         return std::make_pair(lower_bound, upper_bound);
     }
+    /*
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:292]: np: new line between end of function and comment.

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 2]
```diff
@@ -289,6 +289,50 @@ class cursor_bound_01 : public test {
...
+     * Validate the bound search call. If the key is within range, the range cursor should return
+     * back the search key.
+     */
+    void
+    validate_bound_search(int range_ret, scoped_cursor &range_cursor, scoped_cursor &normal_cursor,
+      const std::string &search_key, const bound &lower_bound, const bound &upper_bound)
+    {
+        auto lower_key = lower_bound.get_key();
+        auto upper_key = upper_bound.get_key();
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:301]: np: lets all the variables the moment they are used.

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 3]
```diff
@@ -289,6 +289,50 @@ class cursor_bound_01 : public test {
...
+        testutil_check(range_cursor->get_key(range_cursor.get(), &key));
+        logger::log_msg(LOG_TRACE,
+          "bounded search found key: " + std::string(key) + " with lower bound: " + lower_key +
+            " upper bound: " + upper_key);
+
+        /*
+         * Assert that if range cursor returned with a key, the search key has to be within range.
+         */
+        auto above_lower_key = lower_key.empty() ||
+          custom_lexicographical_compare(lower_key, search_key, lower_inclusive);
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:315]: Instead of search_key, lets test against the returned key from range cursor 🍔

Comment (clarissecheah) [reply test/cppsuite/tests/cursor_bound_01.cpp:315]: I think testing against the search_key would be fine, as that means we can also test the case when range cursor returns WT_NOTFOUND.

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 4]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
...
+         * If bounded cursor returns a valid key, search key must have been in bounds. If normal
+         * cursor returns a valid key, but bounded cursor returns WT_NOTFOUND, the search key must
+         * have been out of bounds.
+         */
+        if (range_ret == 0) {
+            testutil_assert(above_lower_key && below_upper_key);
+            /* Check that the returned key should be equal to the search key. */
+            testutil_assert(std::string(key).compare(search_key) == 0);
+        } else if (range_ret == WT_NOTFOUND) {
+            testutil_assert(!(above_lower_key && below_upper_key));
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:331]: Remove braces for one liners

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 5]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
...
 
+    /*
+     * Validate the bound search call. If the key is within range, the range cursor should return
+     * back the search key.
+     */
+    void
+    validate_bound_search(int range_ret, scoped_cursor &range_cursor, scoped_cursor &normal_cursor,
+      const std::string &search_key, const bound &lower_bound, const bound &upper_bound)
+    {
+
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:301]: extra line

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 6]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
         return std::make_pair(lower_bound, upper_bound);
     }
 
+    /*
+     * Validate the bound search call. If the key is within range, the range cursor should return
+     * back the search key.
+     */
+    void
+    validate_bound_search(int range_ret, scoped_cursor &range_cursor, scoped_cursor &normal_cursor,
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:298]: I think we don't need to pass in the normal cursor anymore 🍔

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 7]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
...
+          "bounded search found key: " + std::string(key) + " with lower bound: " + lower_key +
+            " upper bound: " + upper_key);
+
+        /*
+         * Assert that if range cursor returned with a key, the search key has to be within range.
+         */
+        auto above_lower_key = lower_key.empty() ||
+          custom_lexicographical_compare(lower_key, search_key, lower_inclusive);
+        auto below_upper_key = upper_key.empty() ||
+          custom_lexicographical_compare(search_key, search_key, upper_inclusive);
```
Comment (etienneptl) [inline test/cppsuite/tests/cursor_bound_01.cpp:319]: Seems that we have `search_key` twice here, is it expected?

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 8]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
...
+
+        /*
+         * If bounded cursor returns a valid key, search key must have been in bounds. If normal
+         * cursor returns a valid key, but bounded cursor returns WT_NOTFOUND, the search key must
+         * have been out of bounds.
+         */
+        if (range_ret == 0) {
+            testutil_assert(above_lower_key && below_upper_key);
+            /* Check that the returned key should be equal to the search key. */
+            testutil_assert(std::string(key).compare(search_key) == 0);
```
Comment (etienneptl) [inline test/cppsuite/tests/cursor_bound_01.cpp:329]: np: You can do `search_key.compare(key)` instead so you don't have to cast, see the [docs](https://cplusplus.com/reference/string/string/compare/).

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 9]
```diff
@@ -290,6 +290,48 @@ class cursor_bound_01 : public test {
...
+        } else if (range_ret == WT_NOTFOUND) {
+            testutil_assert(!(above_lower_key && below_upper_key));
+        }
```
Comment (etienneptl) [inline test/cppsuite/tests/cursor_bound_01.cpp:332]: What if `range_ret` has another value? Is it checked anywhere?

File: test/cppsuite/tests/cursor_bound_01.cpp (hunk 8 above)
Comment (clarissecheah) [reply test/cppsuite/tests/cursor_bound_01.cpp:329]: Good idea, thanks!

File: test/cppsuite/tests/cursor_bound_01.cpp (hunk 7 above)
Comment (clarissecheah) [reply test/cppsuite/tests/cursor_bound_01.cpp:319]: Nope. Fixed that!

File: test/cppsuite/tests/cursor_bound_01.cpp (hunk 9 above)
Comment (clarissecheah) [reply test/cppsuite/tests/cursor_bound_01.cpp:332]: It's not been checked, thanks for pointing that out.

File: test/cppsuite/tests/cursor_bound_01.cpp (hunk 6 above)
Comment (clarissecheah) [reply test/cppsuite/tests/cursor_bound_01.cpp:298]: You're right, thanks!

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 10]
```diff
@@ -290,6 +290,47 @@ class cursor_bound_01 : public test {
...
+
+        /*
+         * Assert that if range cursor returned with a key, the search key has to be within range.
+         */
+        auto above_lower_key = lower_key.empty() ||
+          custom_lexicographical_compare(lower_key, search_key, lower_inclusive);
+        auto below_upper_key = upper_key.empty() ||
+          custom_lexicographical_compare(search_key, upper_key, upper_inclusive);
+
+        testutil_assert(range_ret == 0 || range_ret == WT_NOTFOUND);
```
Comment (jiechenbo) [inline test/cppsuite/tests/cursor_bound_01.cpp:320]: We don't need to do this check, we do it before we call the function 😄

**Review by jiechenbo — APPROVED** [review]
LGTM!

File: test/cppsuite/tests/cursor_bound_01.cpp [hunk 11]
```diff
@@ -290,6 +290,47 @@ class cursor_bound_01 : public test {
...
+        /*
+         * If bounded cursor returns a valid key, search key must have been in bounds. If normal
+         * cursor returns a valid key, but bounded cursor returns WT_NOTFOUND, the search key must
+         * have been out of bounds.
+         */
+        if (range_ret == 0) {
+            testutil_assert(above_lower_key && below_upper_key);
+            /* Check that the returned key should be equal to the search key. */
+            testutil_assert(search_key.compare(key) == 0);
+        } else if (range_ret == WT_NOTFOUND)
```
Comment (etienneptl) [inline test/cppsuite/tests/cursor_bound_01.cpp:330]: You don't need to check for the value again here, using "else" should be fine!

File: test/cppsuite/tests/cursor_bound_01.cpp (hunk 10 above)
Comment (jiechenbo) [reply test/cppsuite/tests/cursor_bound_01.cpp:320]: Hmm think we should keep this check and remove the other one

**Review by jiechenbo — APPROVED** [review]
LGTM!

**Pull request merged**